File: cpp/leetcode/396.cpp
class Solution {
public:
    int maxRotateFunction(vector<int>& A) {
        int F = 0, total = 0;
        for(int i = 0; i < A.size(); i++){
            F += i * A[i];
            total += A[i];
        }
        int res = F, n = A.size();
        for(int i = n - 1; i >= 0; i--){
            F = F + total - n * A[i];
            res = max(res, F);
        }
        return res;
    }
};

File: python/leetcode/396.py
class Solution(object):
    def maxRotateFunction(self, nums):
        if not nums: return 0
        res = sum([i * nums[i] for i in xrange(len(nums))])
        total, n, tmp = sum(nums), len(nums), res
        for idx in xrange(-1, - n - 1, -1):
            res = max(res, tmp)
            tmp += total - (n * nums[idx])
        return res